### PR TITLE
Fix bond deletion

### LINKED ISSUE #4762: delete_bonds not deleting all bonds
## Expected behavior ##

If I call `u.delete_bonds(u.bonds)`, I expect all bonds to be deleted.

## Actual behavior ##

This doesn't always happen. If I had to guess where to look, it'd be in `_Connection._delete_bonds` where there's some indexing logic that could be error-prone. 

## Code to reproduce the behavior ##

``` python
>>> import MDAnalysis as mda
>>> from MDAnalysis.tests.datafiles import PDB_conect
>>> u = mda.Universe(PDB_conect)
>>> u.bonds
<TopologyGroup containing 8 bonds>
>>> u.delete_bonds(u.bonds)
>>> u.bonds
<TopologyGroup containing 1 bonds>
>>> u.delete_bonds(u.bonds)  # second time's the charm
>>> u.bonds
<TopologyGroup containing 0 bonds>
```

## Current version of MDAnalysis ##

- Which version are you using? (run `python -c "import MDAnalysis as mda; print(mda.__version__)"`) develop
- Which version of Python (`python -V`)? 3.10.12
- Which operating system? ARM Mac

### PR DESCRIPTION
**Pull request by lilyminium**
Fixes #4762

Changes made in this Pull Request:
 - ~~Changes Topology creation and addition to only accept a single identical bond/angle/dihedral/etc instead of multiple~~ (see discussion in PR)
 - Changes bond/angle/etc deletion to delete all copies of an identical bond if that bond is passed in for deletion

PR Checklist
------------
 - [x] Tests?
 - [ ] Docs?
 - [x] CHANGELOG updated?
 - [x] Issue raised/referenced?

## Developers certificate of origin
- [x] I certify that this contribution is covered by the LGPLv2.1+ license as defined in our [LICENSE](https://github.com/MDAnalysis/mdanalysis/blob/develop/LICENSE) and adheres to the [**Developer Certificate of Origin**](https://developercertificate.org/).

----
📚 Documentation preview 📚: https://mdanalysis--4763.org.readthedocs.build/en/4763/

### CODE REVIEW

File: testsuite/MDAnalysisTests/topology/test_pdb.py [hunk 1]
```diff
@@ -146,7 +146,7 @@ def parse():
         struc = parse()
 
     assert hasattr(struc, 'bonds')
-    assert len(struc.bonds.values) == 4
+    assert len(struc.bonds.values) == 2
```
Comment (lilyminium) [inline testsuite/MDAnalysisTests/topology/test_pdb.py]: These used to be bonds between [(0, 1), (0, 1), (1, 2), (1, 2)] -- so two bonds repeated. This change makes it so each bond is unique.

File: testsuite/MDAnalysisTests/topology/test_pdb.py [hunk 2]
```diff
@@ -158,7 +158,7 @@ def parse():
     with pytest.warns(UserWarning):
         struc = parse()
     assert hasattr(struc, 'bonds')
-    assert len(struc.bonds.values) == 2
+    assert len(struc.bonds.values) == 1
```
Comment (lilyminium) [inline testsuite/MDAnalysisTests/topology/test_pdb.py]: Similar to above.

**Comment by IAlibay** [comment]
@lilyminium this PR is marked as draft - is it meant to be up for review or is there more work that needs to be done?

**Review by IAlibay — COMMENTED** [review]
Overall this looks good to me - however there's one likely outcome of this "behaviour change" that might have a real impact.

Some software intentionally misuse the PDB format's CONECT records to represent different bond orders - PyMol is a good example of this iirc. In those cases, my understanding is that with this change you would lose that information?

As far as I'm aware, the PDB writer would lose that information even if you have duplicate bonds because we call a set on the bonds of all the atoms (which should squash out duplicates?). So there's probably a good case to be made that support for multi-conect PDBs is not something we want to do anyways.

I'm not particularly sure that I care about the behaviour change, but I do want to cc @MDAnalysis/coredevs (especially @orbeckst @yuxuanzhuang and @RMeli ) who might have some more immediate thoughts on the whole thing.

**Comment by IAlibay** [comment]
Failures aren't happening in other PRs - it looks like something that has been specifically introduced here.

File: package/MDAnalysis/core/topologyattrs.py [hunk 3]
```diff
@@ -3118,9 +3111,8 @@ def _add_bonds(self, values, types=None, guessed=True, order=None):
         if order is None:
             order = itertools.cycle((None,))
 
-        existing = set(self.values)
         for v, t, g, o in zip(values, types, guessed, order):
-            if v not in existing:
+            if v not in self.values:
```
Comment (IAlibay) [inline package/MDAnalysis/core/topologyattrs.py]: This change is likely the culprit - doing a `not in` operation on a list is orders of magnitude slower than on a set.

Example:

```python
In [1]: import random

In [2]: lister = [random.random() for i in range(10000)]

In [3]: setter = set(lister)

In [4]: len(setter)
Out[4]: 10000

In [5]: len(lister)
Out[5]: 10000

In [6]: def notin(a):
   ...:     for v in range(10000):
   ...:         if v not in a:
   ...:             pass
   ...:     return
   ...: 

In [7]: %timeit notin(lister)
734 ms ± 5.85 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [8]: %timeit notin(setter)
129 μs ± 166 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
```

**Review by IAlibay — CHANGES_REQUESTED** [review]
We should try to revert to a set for the `not in` comparison, or do some other kind of trick to avoid the performance loss.

**Comment by lilyminium** [comment]
Hmm. Yes that would be the case that multi-bonds can't be sketchily represented by digging into Topology internals, which downstream people may well be doing.

I can think of two solutions to this in this PR:

* put that information in the `order` attribute (which TopologyObjects should have) -- a bit of work, may have time later today to push it through but not guaranteed. Also, even more changes in current behaviour
* This PR fixes the deletion issue twice over. It first only counts identical bonds once when creating /adding to the topology object. It secondly also deletes all identical bonds when we try to delete. If we take out the first and just ensure that deletion works, we should still fix the original issue. This is a lot less work

**Comment by lilyminium** [comment]
Pushed a commit that removes the "addition" fixes and preserves previous behaviour (should be easy to revert if discussion goes the other way)

**Comment by lilyminium** [comment]
I'm leaning towards the simpler fix of just ensuring that all identical bonds are deleted when deletion is caused, as it's less disruptive to current behaviour, and potentially fixing the addition of identical bonds in a separate issue/PR later, so I've updated the changelog too to reflect that -- should still be easy to revert if there's any disagreement

File: package/MDAnalysis/core/topologyattrs.py (hunk 3 above)
Comment (lilyminium) [reply package/MDAnalysis/core/topologyattrs.py]: I removed this in the commit below, but note to self that changing `self.values` to a dict might be the easiest way to do this performantly and retain both uniqueness and order.

Or masking the input `values` after checking a `set` version of `self.values` might be less work but still more performant.

**Review by IAlibay — APPROVED** [review]

**Comment by IAlibay** [comment]
Going with just the bond deletion fix seems like a reasonable thing for this release.

**Pull request merged**